pkg/syncer/volume_health_reconciler.go

```diff
@@ -153,10 +254,26 @@ func (rc *volumeHealthReconciler) Run(
 	stopCh := ctx.Done()
 
 	if !cache.WaitForCacheSync(stopCh, rc.tkgPVSynced, rc.svcPVCSynced) {
-		log.Errorf("Cannot sync supervisor and/or TKG pv/pvc caches")
+		log.Error("Cannot sync supervisor and/or TKG pv/pvc caches")
 		return
 	}
 
+	// Initialize volumeHandleToPVs map with existing PV's
+	// Since this mapping is lost across pvCSI restarts, it needs to be
+	// initialized with existing values to prevent missing an SVC-PVC
+	// event due to absence of an entry in the map.
+	pvs, err := rc.tkgPVLister.List(labels.Everything())
+	if err != nil {
+		log.Errorf("failed to list TKG PV's with error: %+v", err)
+		return
```

Comment:
Looks like we will never run volume health reconciler if we fail to list the PVs. If so, can you fix it?

Reply:
I've already raised an issue for this - https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/411
The same exists if `WaitForCacheSync` fails either here or in resize reconciler. 
I plan to fix this in a follow up PR

Reply:
For the code you are adding why not fix it now? Wouldn't it be just a matter of moving this code to NewVolumeHealthReconciler (before adding event handlers to svcPVCInformer)?

Reply:
I didn't do that for two reasons:
1. Currently `NewVolumeHealthReconciler` does not return an error, and is called within a `OnceFor...` scope - https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/syncer/metadatasyncer.go#L1151
That `OnceFor..` was added as part of a change to handle FSS config map changes, to ensure that the reconciler is only initialized once. I'd need to look more into whether that's needed or not and how to modify it without breaking functionality. 
2. Since the informers and caches are not set up in `NewVolumeHealthReconciler`, i'd have to directly query the API server, instead of listing it from the cache (which is what i'm doing here). 

But i hear the concern. Changing this log to `log.Fatalf` which, according to the docs:
```
// Fatalf uses fmt.Sprintf to log a templated message, then calls os.Exit.
``` 
So the container will restart.

Reply:
Also fixing https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/411 as part of this change since its a one liner..

Reply:
I think this is not the right way to handle this. Restarting containers will affect other logic we have.

Reply:
That's the way this is handled by external-provisioner - https://github.com/kubernetes-csi/external-provisioner/blob/master/cmd/csi-provisioner/csi-provisioner.go#L347

Reply:
The current design principle we are following is to not restart the containers if we are unable to connect to another service. We have handled VC connection issues by failing the workflow and reestablishing the connection on retry. For k8s client connection too, we need to be doing this same. This will help us to not throw away the cache we have built, or trigger expensive workflows like full sync, etc. Also many customers see CSI container restarts as problematic. Since this PR is going against this principle and crashing the container, we should fix it.

Reply:
This is not the principle we follow today. If metadata syncer cant create a k8sclient (https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/syncer/metadatasyncer.go#L121) or connect to the vCenter (https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/syncer/metadatasyncer.go#L152) during initialization then we return an error and call `os.Exit()` (https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/cmd/syncer/main.go#L141) which restarts the container. 
Same if Cns Operator (https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/cmd/syncer/main.go#L133) cant connect to the vCenter.

Reply:
This is not just in syncer. If CSI controller can't connect to the vCenter, we return an error and the container restarts - https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/csi/service/vanilla/controller.go#L126

Reply:
All the links you pointed to crashes the CSI container during startup. Do we crash the container during runtime after successful init?

Reply:
`Run()` is currently called from `initVolumeHealthReconciler` so i'm not sure we can say that volume health reconciler is ever successfully initialized. 
I can move this piece of code into `NewVolumeHealthReconciler` with the caveat that we're listing from API server rather than cache (although its a one time cost). 
For the `WaitForCacheSync` issue, i will address that in another PR.